### PR TITLE
Expose GDScript syntax highlighter to editor plugins

### LINKED ISSUE #10374: Expose GDScript syntax highlighter to editor plugins
### Describe the project you are working on

A components plugin that makes it easy to selectively add external scripts to the project

### Describe the problem or limitation you are having in your project

I made a quick preview popup that shows source code before you add a script to the project but it doesn't have syntax highlighting (and it would help if it did):

gc

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

A reference to GDScript syntax highlighter that can be used in `TextEdit.syntax_highlighter`

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Could be a static property on `EditorSyntaxHighlighter` or maybe a method on the `EditorInterface` to get it, or just the class itself instead of an instance. I'm not sure tbh, maybe knowing how it works internally will help deciding

### If this enhancement will not be used often, can it be worked around with a few lines of script?

`EditorInterface.get_script_editor().get_current_editor().get_base_editor().syntax_highlighter` but it relies on some script being opened in the script editor. I thought about opening some dummy script but as far as I can tell there's no way to do it when using external editor. Even if it was possible it's way too hacky

### Is there a reason why this should be core and not an add-on in the asset library?

Unavailable to addons

### PR DESCRIPTION
**Pull request by passivestar**
Closes https://github.com/godotengine/godot-proposals/issues/10374

| Before | After |
|--------|--------|
| before | after |

### CODE REVIEW

File: modules/gdscript/doc_classes/GDScriptSyntaxHighlighter.xml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="GDScriptSyntaxHighlighter" inherits="EditorSyntaxHighlighter" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../doc/class.xsd">
+	<brief_description>
+		GDScript syntax highlighter class.
```
Comment (dalexeev) [inline modules/gdscript/doc_classes/GDScriptSyntaxHighlighter.xml:4]: I think the description can be improved, please see [the docs](https://docs.godotengine.org/en/stable/contributing/documentation/docs_writing_guidelines.html#give-an-overview-of-the-node-in-the-brief-description).

File: modules/gdscript/doc_classes/GDScriptSyntaxHighlighter.xml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="GDScriptSyntaxHighlighter" inherits="EditorSyntaxHighlighter" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../doc/class.xsd">
+	<brief_description>
+		A GDScript syntax highlighter that can be used with [TextEdit] and [CodeEdit] nodes
+	</brief_description>
+	<description>
+		[b]Note:[/b] This class can only be used for editor plugins because it relies on editor settings.
+		[b]Example:[/b]
```
Comment (AThousandShips) [inline modules/gdscript/doc_classes/GDScriptSyntaxHighlighter.xml:8]:
```suggestion
```
As per:
* https://github.com/godotengine/godot/pull/95749

File: modules/gdscript/doc_classes/GDScriptSyntaxHighlighter.xml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="GDScriptSyntaxHighlighter" inherits="EditorSyntaxHighlighter" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../doc/class.xsd">
+	<brief_description>
+		A GDScript syntax highlighter that can be used with [TextEdit] and [CodeEdit] nodes
```
Comment (dalexeev) [inline modules/gdscript/doc_classes/GDScriptSyntaxHighlighter.xml:4]:
```suggestion
		A GDScript syntax highlighter that can be used with [TextEdit] and [CodeEdit] nodes.
```

File: modules/gdscript/doc_classes/GDScriptSyntaxHighlighter.xml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+	</brief_description>
+	<description>
+		[b]Note:[/b] This class can only be used for editor plugins because it relies on editor settings.
+		[codeblocks]
+		[gdscript]
+		var code_preview = TextEdit.new()
+		var highlighter = GDScriptSyntaxHighlighter.new()
+		code_preview.syntax_highlighter = highlighter
+		[/gdscript]
+		[/codeblocks]
```
Comment (dalexeev) [inline modules/gdscript/doc_classes/GDScriptSyntaxHighlighter.xml:14]:
```suggestion
		[codeblock]
		var code_preview = TextEdit.new()
		var highlighter = GDScriptSyntaxHighlighter.new()
		code_preview.syntax_highlighter = highlighter
		[/codeblock]
```

Or alternatively you can add `[csharp][/csharp]`.

File: modules/gdscript/doc_classes/GDScriptSyntaxHighlighter.xml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+	</description>
+	<tutorials>
+	</tutorials>
```
Comment (dalexeev) [inline modules/gdscript/doc_classes/GDScriptSyntaxHighlighter.xml:17]: No need to change, I just noticed that for some reason we require an empty `<tutorials></tutorials>` even though for the other member types we remove empty groups.

**Review by dalexeev — COMMENTED** [review]
It would be nice to add `text_edit` checks before exposing the class. It's unlikely that anyone will call `get_line_syntax_highlighting()` directly, but still.

https://github.com/godotengine/godot/blob/7a4a6fbc039fefbccbb23a66285722ad20bf418a/modules/gdscript/editor/gdscript_highlighter.cpp#L96

https://github.com/godotengine/godot/blob/7a4a6fbc039fefbccbb23a66285722ad20bf418a/modules/gdscript/editor/gdscript_highlighter.cpp#L693

**Comment by passivestar** [comment]
@dalexeev

> It would be nice to add text_edit checks

It doesn't look like we need to. The base class already does it in both `get_line_syntax_highlighting()` and `update_cache()` and they aren't virtual

https://github.com/godotengine/godot/blob/7a4a6fbc039fefbccbb23a66285722ad20bf418a/scene/resources/syntax_highlighter.cpp#L42-L44

https://github.com/godotengine/godot/blob/7a4a6fbc039fefbccbb23a66285722ad20bf418a/scene/resources/syntax_highlighter.cpp#L79-L81

**Review by dalexeev — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**